os_info/src/windows/winapi.rs

```diff
@@ -114,17 +122,80 @@ fn version_info() -> Option<OSVERSIONINFOEX> {
     }
 }
 
+fn str_to_wide(value: &str) -> Vec<WCHAR> {
+    OsStr::new(value).encode_wide().chain(Some(0)).collect()
+}
+
+fn release_id() -> Option<String> {
+    const REG_SUCCESS: LSTATUS = ERROR_SUCCESS as LSTATUS;
+    const MAX_CHARS: usize = 256;
+    // Assume something will go wrong
+    let mut rv = None;
+    // Path as a wide string (array of WCHAR)
+    let sub_key = str_to_wide("SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion");
+    // Registry key
+    let mut key: HKEY = std::ptr::null_mut();
+    // Try to open the path for reading (should never fail)
+    if unsafe { RegOpenKeyExW(HKEY_LOCAL_MACHINE, sub_key.as_ptr(), 0, KEY_READ, &mut key) }
+        == REG_SUCCESS
+    {
+        // Name as a wide string
+        let name = str_to_wide("ReleaseId");
+        // Where we'll be storing
+        let mut data_type: DWORD = 0;
+        let mut data = [0 as WCHAR; MAX_CHARS];
+        let mut data_size: DWORD = std::mem::size_of::<[WCHAR; MAX_CHARS]>() as DWORD;
+        // Try reading the value
+        let status = unsafe {
+            RegQueryValueExW(
+                key,
+                name.as_ptr(),
+                std::ptr::null_mut(),
+                &mut data_type,
+                &mut data as *mut _ as LPBYTE,
+                &mut data_size,
+            )
+        };
+        // Success and the correct datatype?
+        if status == REG_SUCCESS && data_type == REG_SZ {
+            // RegQueryValueExW returns the number of bytes stored including the null terminator.
+            // We want the string length without the terminator.
+            let string_length = (data_size as usize / std::mem::size_of::<WCHAR>()) - 1;
+            let as_slice =
+                unsafe { std::slice::from_raw_parts(&data as *const WCHAR, string_length) };
+            // Convert to a dirty UTF-8 string
+            let osstr: OsString = OsStringExt::from_wide(as_slice);
+            // Drop invalid characters and ensure we own it.  Finally.  A clean Rust string.
+            let value = osstr.to_string_lossy().to_string();
+            rv = Some(value);
+        }
+    }
+    rv
+}
+
 // Examines data in the OSVERSIONINFOEX structure to determine the Windows edition:
 // https://msdn.microsoft.com/en-us/library/windows/desktop/ms724833(v=vs.85).aspx
-fn edition(version_info: &OSVERSIONINFOEX) -> Option<String> {
+fn edition<R>(version_info: &OSVERSIONINFOEX, release_id: R) -> Option<String>
```

Comment:
Do I understand correctly that you are using the closure here only to be able to test this function? In that case I think that it would be better to move this logic outside of the `edition` function. Something like that:
```rust
let edition = match edition(...) {
    Some("Windows Server 2016") => {
        if let Some(rid) = release_id() {
            Some("Windows Server 2019")
        }
    }
    e => e,
}
```

Reply:
Hello @stanislav-tkach!

> ...more convenient to check the dwBuildNumber field of the OSVERSIONINFOEX structure...

Oh, I agree!

I went with release ID because Microsoft has made a public commitment that...
* It's a code date so any given value will always be greater than the previous value.
* They will always coincide with a major upgrade.

While I don't believe Microsoft has ever reset `dwBuildNumber`, as far as I know, they've never committed to that.  So, it is theoretically possible they could increment the minor version and reset the build number.  Given the structure of the `match` expression in `edition` it's not a problem.

But, this is an opportunity to consider another option.  The `ProductName` value from that same Registry path is worth considering.  For example, on the Windows 10 computer in front of me the value is `Windows 10 Pro` and the value from an AWS 2016 instance is `Windows Server 2016 Datacenter`.  Reading that value, if it's available, would save us from the constant nag of having to keep `edition` up-to-date and `edition` would return a bit more detail than it currently does.  Thoughts?

Reply:
It's also to eliminate the overhead for those who don't need to fetch the value from the Registry.  Currently that would be Windows 10 users.

Moving it outside is also a good choice.  I personally don't have a strong opinion one way or the other.

I'll play with it a bit then let you know if I develop an opinion.

Reply:
I agree that reading registry can give more precise results. Additionally I have just thought that if Windows Server 2016 is updated, then build matching over the build number will be wrong.